gherkin-utils/java/src/main/java/io/cucumber/gherkin/util/PrettyPrintGherkinDocument.java

```diff
@@ -0,0 +1,371 @@
+                    .concat("\n");
+        }
+
+        @Override
+        public String handleExamples(Examples examples, String content) {
+            List<TableRow> tableRows = new ArrayList<>();
+            if (examples.getTableHeader() != null) {
+                tableRows.add(examples.getTableHeader());
+                tableRows.addAll(examples.getTableBody());
+            }
+            return content
+                    .concat(prettyKeywordContainer(examples, syntax, scenarioLevel + 1))
+                    .concat(prettyTableRows(tableRows, syntax, scenarioLevel + 2));
+        }
+
+        @Override
+        public String handleRule(Rule rule, String content) {
+            scenarioLevel = 2;
+            return content.concat(prettyKeywordContainer(rule, syntax, 1));
+        }
+
+        @Override
+        public String handleScenario(Scenario scenario, String content) {
+            return content.concat(prettyKeywordContainer(scenario, syntax, scenarioLevel));
+        }
+
+        @Override
+        public String handleStep(Step step, String content) {
+            return content
+                    .concat(stepPrefix(scenarioLevel + 1, syntax))
+                    .concat(step.getKeyword())
+                    .concat(step.getText())
+                    .concat("\n");
+        }
+    }
+
+    /**
+     * Pretty print a GherkinDocument as either markdown or gherkin.
+     *
+     * @param gherkinDocument The GherkinDocument object.
+     * @param syntax          What syntax do you want to format to?
+     * @return The pretty printed string representation of the Gherkin doc in the syntax indicated.
+     */
+    public static String prettyPrint(GherkinDocument gherkinDocument, Syntax syntax) {
+        return WalkGherkinDocument.walkGherkinDocument(gherkinDocument, "",
+                new PrettyPrintGherkinDocumentHandlers(syntax));
+    }
+
+    private static String prettyLanguageHeader(String language) {
+        return "en".equals(language) ? "" : String.format("# language: %s\n", language);
+    }
+
+    private static String prettyKeywordContainer(
+            Scenario stepContainer,
+            Syntax syntax,
+            int level) {
+        List<Tag> tags = stepContainer.getTags() != null ? stepContainer.getTags() : Collections.emptyList();
+        int stepCount = stepContainer.getSteps() != null ? stepContainer.getSteps().size() : 0;
+        String description = prettyDescription(stepContainer.getDescription(), syntax);
+        return (level == 0 ? "" : "\n") +
+                prettyTags(tags, syntax, level) +
+                keywordPrefix(level, syntax) +
+                stepContainer.getKeyword() +
+                ": " +
+                stepContainer.getName() +
+                "\n" +
+                description +
+                (description != null && !"".equals(description.trim()) && stepCount > 0 ? "\n" : "");
+    }
+
+    private static String prettyKeywordContainer(
```

Comment:
There is a lot of duplication in this method and the next three. Can you extract a private method that they all use?